ansible_base/lib/utils/views.py

```diff
@@ -0,0 +1,82 @@
+
+from ansible_base.lib.utils.settings import get_function_from_setting, get_setting
+
+logger = logging.getLogger('ansible_base.lib.utils.views')
+
+
+class AnsibleBaseView(APIView):
+    def initialize_request(self, request, *args, **kwargs):
+        """
+        Store the Django REST Framework Request object as an attribute on the
+        normal Django request, store time the request started.
+        """
+        self.time_started = time.time()
+
+        # If there are any custom headers in REMOTE_HOST_HEADERS, make sure
+        # they respect the allowed proxy list
+        proxy_ip_allowed_list = get_setting('PROXY_IP_ALLOWED_LIST', None)
+        if proxy_ip_allowed_list and all(
+            [
+                request.environ.get('REMOTE_ADDR') not in proxy_ip_allowed_list,
+                request.environ.get('REMOTE_HOST') not in proxy_ip_allowed_list,
+            ]
+        ):
+            for custom_header in get_setting('REMOTE_HOST_HEADERS', None):
+                if custom_header.startswith('HTTP_'):
+                    request.environ.pop(custom_header, None)
+
+        return super().initialize_request(request, *args, **kwargs)
+
+    def finalize_response(self, request, response, *args, **kwargs):
+        response = super().finalize_response(request, response, *args, **kwargs)
+
+        if request.user and request.user.is_authenticated:
+            version = 'Unknown'
+            the_function = get_function_from_setting('ANSIBLE_BASE_PRODUCT_VERSION_FUNCTION', None)
+            if the_function:
+                try:
+                    version = the_function()
+                except Exception:
+                    logger.exception(_('ANSIBLE_BASE_PRODUCT_VERSION_FUNCTION was set but calling it as a function failed (see exception).'))
+
+            response['X-API-Product-Version'] = version
+
+        response['X-API-Product-Name'] = get_setting('ANSIBLE_BASE_PRODUCT_NAME', _('Unnamed'))
+        response['X-API-Node'] = get_setting('CLUSTER_HOST_ID', _('Unknown'))
+
+        time_started = getattr(self, 'time_started', None)
+        if time_started:
+            time_elapsed = time.time() - self.time_started
+            response['X-API-Time'] = '%0.3fs' % time_elapsed
+
+        if getattr(self, 'deprecated', False):
+            response['Warning'] = _('This resource has been deprecated and will be removed in a future release.')
+
+        return response
+
+
+# Determine and load the parent view
+# If anything fails in here its a pretty low level exception so we don't catch anything and just let it raise
+parent_view = getattr(settings, 'ANSIBLE_BASE_CUSTOM_VIEW_PARENT', None)
+parent_view_class = AnsibleBaseView
+if parent_view:
+    try:
+        module_name, junk, class_name = parent_view.rpartition('.')
+        logger.debug(f"Trying to import parent view {class_name} from package {module_name}")
+        module = importlib.import_module(module_name, package=class_name)
```

Comment:
This isn't circular?

Say I set `ANSIBLE_BASE_CUSTOM_VIEW_PARENT`, and that should be a class that inherits from `AnsibleBaseView`. Python imports my class, that imports this module, now on line 73 it imports my class again.

If that is a problem, or just to make it less circular, I could see it making sense to put ln 65 and after in another module. That way, when it imports my class, my class imports `AnsibleBaseView` and fully completes the import of the entire module its in and can return control here without additional smarts.

Reply:
You'd think it would be but we can setup a quick test does not show a circular import. Lets meet up tomorrow to discuss. The brain is too tired tonight :)

Reply:
Split into seperate files.